### ISSUE TITLE
Query support implementation on Android

### ISSUE THREAD

**Issue by cyrilnoah1** [oltrenuovefrontiere/fluttercouch#11]
**Is your feature request related to a problem? Please describe.**
Couchbase query processing implementation on Android.

**Describe the solution you'd like**
My solution for the implementation is to let the user create a Query object and append all the query related clause as functions, and finally call run() or runLive() to get the according result. 
Here is a sample of the solution I suggest:

```dart
_runQuery() async {
Query myQuery = Query();
List<Employee> employees = await myQuery.select(Query.ALL, Employee::class)
.where("_id", Query.EQUALS, empId /*user provided employee id*/)
.and("employee.salary", Query.GREATER_THAN, 10_000_000L)
.and("employee.experience", Query.LESS_THAN, 5 /*years*/)
.run(); // or runLive()
}
```

**Comment by cyrilnoah1** [comment]
Hi, @oltrenuovefrontiere. Firstly, I wanted to thank you again for adding me as a contributor to the project. I wanted to let you know that I'm still a newbie to Flutter and Dart, hence I have to tighten my grip on trip on them. However, here's what I propose the final representation to look like:

```dart
class Employee{
String _id;
double salary;
float experience;

Employee(this._id, this.salary, this.experience);

static const ID = "_id";
static const SALARY = "salary";
static const EXPERIENCE = "experience";
}

// Running a query async.
_runQuery() async{
List<Employee> employees = await myQuery.select(Query.ALL, Employee::class)
.where(Employee.ID, Query.EQUALS, empId /*user provided employee id*/)
.and(Employee.SALARY, Query.GREATER_THAN, 10_000_000L)
.and(Employee.EXPERIENCE, Query.LESS_THAN, 5 /*years*/)
.run(); // or runLive()
}
```

But, there's a caveat in implementing the above representation. Notice that we are not using any callbacks to return the data when calling run. To achieve direct returning of the result, my proposal is to use Kotlin Coroutines on the Android native front, where we can use the **suspendCoroutine** to convert our internal callbacks into a single nullable object (Not sure if we can achieve the same with RxJava/RxAndroid). Here's what it would look like:

```kotlin
fun run() = suspendCoroutine<T?>{
// Our query with the final result callbacks.
queryBuilder.result(/*model related parameters*/, object: Query.QueryResult{

override fun onSuccess(model: T){
it.resume(model)
}

override fun onFailure(){
it.resume(null)
}
})
}
```

But, to achieve this, we might have to write bits of Android native in Kotlin, please let me know if that's fine by you. 

Please provide your opinion on my proposal.

Thank you,
Cyril

**Comment by oltrenuovefrontiere** [comment]
Thanks for sharing your intuitions about implementing Queries in Fluttercouch. It’s an extremely attractive feature and it will advantage the package quality!
Your proposition is certainly like Couchbase Lite native libraries and is doubtless the form we finally realize it.

To just update you about our earlier ideas, I part with you two approaches we could use for queries differently. (based upon my explorations and exchange of views with @sensiblearts). 
The purpose is that since those two are faster to complete and test, we employ them as quicker step to achieve the objective, while we will start your proposal-like path.
**Feel free to express your point of view on that. I drafted this merely to catch you up about earlier thinking about the question and we haven’t then established a precise roadmap** 

## Native Objects for Queries
It’s feasible to join native library code (which lives in the fluttercouch directory) and application native code. We let the user define their queries in Java/Kotlin and Objective C/Swift, enclose them in an Object, inject into the CBManager, instantiated and called from Dart code through dart<->native channels.
Here a simplistic example for Android:
```java
class MyQueries implements FlutterCouchQueryInterface {
  // Interface method implementation
  public getQueryByName(String name) {
    // here the user specify his queries in native code and return it so CBmanager can control it
   }
}
```

## The String /Javascript approach
Queries can be “inflated” from a string definition, which in CB Lite 1.x(1) should be written in Javascript and in the just-released CB Lite 2.5 in N1ql (an SQL-similar code adopted in Couchbase Server to represent queries). That string can be passed from dart to native code through platform-channel and used to inflate the query, with no intervention by the end user. The major weakness is the complete loss of compiler hints for query building.

## The result set problem
Any concept we adopt for setting up the query, the result set needs to be serialized in native code, passed through platform-channel and deserialized in Dart.
If I‘m not misread you imply reflection in your outline (the Employee::class reference). In Flutter we use code generators instead of reflection (2). In Fluttercouch, we could use platform-channel default deserialiser, which returns dart basic types (int, string, map, list, etc) and investigate a means to efficiently define deserializer methods for custom classes. Otherwise we could serialize the result set as json and let the user define deserialization through his preferred json package.


(1) We desire to support Couchbase Lite 1.x because 2.x lacks replication support for CouchDB. (see fluttercouchDB and issues from @sensiblearts for further reference)

(2) The reason is that dart compiler uses a ‘tree-shaking’ technique and rejects any code not finally employed in that specific application. Reflection is conflicting with this technique because it would be impossible to figure out at compile time which code you can reference at runtime through reflection.

**Comment by oltrenuovefrontiere** [comment]
About your last proposal, there is no problem to use Kotlin for Android development, but the entire project has to be translated to it because we are still using Java...

**Comment by cyrilnoah1** [comment]
Hi, Luca. Thank you for elaborating the pros and cons about my proposal, and describing the underlying functionality. I wasn't aware of that we use code generators in Dart, and hence your approach seems to be a lot more achievable without hurdles in between. 

I guess I will be getting the full gist of it once I learn about the Dart<->native communication mechanism. 

Finally, I would like to say that I completely side with your approach to the implementation, and my proposal can be later added as sugar to the existing code. :)

Thank you,
Cyril

**Comment by oltrenuovefrontiere** [comment]
Hey Cyril,
no problem with late answers, I know we live all in different timezones. I wrote you before going to sleep and I wake up this very morning with your reply, so don't worry!

How would you feel about writing the Native Objects approach in native code while you deep dive in the Flutter ecosystem? I could complete with the dart part if you will be not comfortable with Flutter.

You can reference to this [page](https://docs.couchbase.com/couchbase-lite/2.5/java.html) as documentation for Android CB Lite 2.5 SDK, and to [this](https://flutter.dev/docs/development/platform-integration/platform-channels) for dart<->native communication.

Please let me know if there’s anything I can help you out with.

Thank you,
Luca

**Comment by cyrilnoah1** [comment]
Hi, Luca, thanks for understanding. And, sure, I can work on the native part, but I'm open to help on the Dart part as well, once I get the full grasp, it would also be good experience for me to learn more about Dart and Flutter. It's been a while since I had gone through the CB documentation, so I'll start with the link you've provided, and also look into the Dart<->native communication bridge.

**Comment by bawelter** [comment]
I already have this implemented with query change listeners.  Just working on permission to release to open source.

**Comment by oltrenuovefrontiere** [comment]
Hi Bawalter,
thanks you for showing interest in this project! Any help will be awesome and it seems like you have some experience with Flutter and Couchbase already. I waited to reply to see if you had some last news about permission on your code, but I finally decided to greet you anyway!
We're really looking forward to benefit from your contribution, in any form it will be!
I think Cyril is going to work actively on this enhancement starting from next weeks because is a bit hard-pulled at the moment.

Please let me know if you have any request or suggestion!
Luca

**Comment by bawelter** [comment]
I am waiting on my boss at the moment, he should be getting back from his vacation today or tomorrow and I will be able to ask him on it then.

**Comment by bawelter** [comment]
@oltrenuovefrontiere I got the go ahead for this so I will try to get the code up ASAP.

**Comment by oltrenuovefrontiere** [comment]
Hey what a great news! I'm really looking forwarding to your subsission, to undarstand your approach and help you integrating it in Fluttercouch!
Thank for your commitment!


Ear we soon!

**Comment by bawelter** [comment]
@oltrenuovefrontiere I uploaded all my changes to my forked repo.  Unfortunately I didn't version control the changes in the forked repo originally so I just committed the changes in the master branch and merged the changes you made.

**Comment by oltrenuovefrontiere** [comment]
@bawelter I rapidly looked through your code and it seems an awesome completion of the query implementation I drafted some time ago. I’ll try in these days his functionality but I’m pretty confident it would just work gracefully.
How do you feel about starting a pull request from your master to fluttercouch/queryImplementation branch?

Thank you for showing interest in the future of this project and thanks your boss for disposability, from me and from the entire Flutter and Couchbase community.

**Comment by bawelter** [comment]
@oltrenuovefrontiere thank you, I will be consolidating all query events into one event channel because I don't have the best understanding of how memory is managed with event channels and didn't want any risk of memory leaks.  After this I will start a pull request.

**Comment by oltrenuovefrontiere** [comment]
@bawelter and @cyrilnoah1  I was thinking about providing the possibility to receive from platform channels a raw JSON version of documents and result sets. That way we could facilitate deserialization to custom classes through standard JSON libraries available in the dart ecosystem. What are your thoughs about that?

**Comment by bawelter** [comment]
@oltrenuovefrontiere I don't think it would hurt anything.  I use built values with code generation in my projects if that would work.

Reference: https://github.com/google/built_value.dart

**Comment by cyrilnoah1** [comment]
> @bawelter and @cyrilnoah1 I was thinking about providing the possibility to receive from platform channels a raw JSON version of documents and result sets. That way we could facilitate deserialization to custom classes through standard JSON libraries available in the dart ecosystem. What are your thoughs about that?

Hi, @oltrenuovefrontiere  and @bawelter. Firstly, my apologies for such a delayed reply. I have completely been occupied with an office project, on weekends as well, and hence have not been able to look into the thread for a while. 

Coming to the suggestion, it is a great idea, @oltrenuovefrontiere, it would make it consistent across other platform implementations as well.

**Comment by oltrenuovefrontiere** [comment]
Thank you for your suggestions. I'll move this proposal to a new issue.

I'm looking forward to complete the integration of @bawelter implementation, and then we can move further for this new feature.

**Comment by bawelter** [comment]
@oltrenuovefrontiere the integration is done and just needs reviewed by someone.

**Comment by oltrenuovefrontiere** [comment]
Merged pull request on master.

**Comment by richard457** [comment]
I am trying to use your code ``` query = QueryBuilder.select([
        SelectResult.expression(Meta.id.from("cells")).as("id"),
//        SelectResult.expression(Expression.property("title").from("cells")),
//        SelectResult.expression(
//            Expression.property("description").from("cells")),
//        SelectResult.expression(Expression.property("createdAt").from("cells")),
//        SelectResult.expression(Expression.property("deletedAt").from("cells")),
      ])
          .from(_databaseName, as: "cells")
          .where(Meta.id.from("cells").equalTo(Expression.string('006')));``` where I have a document with 006 in bucket db but Can not get it working like ``` query.addChangeListener((change) => {
            if (change.results != null)
              {
                print(change.results)
                //resultsCallback(change.results);
              }
          });``` does not return anthing when an update is done on other side I am doing something wrong Please help me.
